src/encoded/static/components/navigation/components/SearchBar.js

```diff
@@ -28,71 +28,112 @@ export class SearchBar extends React.PureComponent{
+    static isBrowseOrSearchPage(href) {
+        return href && typeof href === 'string' && (href.indexOf('/browse/') > -1 || href.indexOf('/search/') > -1);
+    }
+
+    constructor(props) {
         super(props);
-        this.toggleSearchAllItems   = this.toggleSearchAllItems.bind(this);
-        this.onSearchInputChange    = this.onSearchInputChange.bind(this);
-        this.onResetSearch          = this.onResetSearch.bind(this);
-        this.onSearchInputBlur      = this.onSearchInputBlur.bind(this);
+        this.toggleSearchAllItems = this.toggleSearchAllItems.bind(this);
+        this.onSearchInputChange = this.onSearchInputChange.bind(this);
+        this.onResetSearch = this.onResetSearch.bind(this);
+        this.onSearchInputBlur = this.onSearchInputBlur.bind(this);
         this.selectItemTypeDropdown = this.selectItemTypeDropdown.bind(this);
+        this.getDropdownTitleText = this.getDropdownTitleText.bind(this);
+        this.getActionType = this.getActionType.bind(this);
 
         var initialQuery = '';
-        if (props.href){
+        if (props.href) {
             initialQuery = searchFilters.searchQueryStringFromHref(props.href) || '';
         }
         this.state = {
-            'searchAllItems'    : props.href && navigate.isSearchHref(props.href),
-            'typedSearchQuery'  : initialQuery
+            'searchItemType': props.href && navigate.isSearchHref(props.href) ? 'all' : 'sets',
+            'typedSearchQuery': initialQuery,
         };
     }
 
-    componentDidUpdate(pastProps){
+    componentDidUpdate(pastProps) {
         const { href } = this.props;
-        if (pastProps.href !== href){
+        if (pastProps.href !== href) {
             const query = searchFilters.searchQueryStringFromHref(href) || '';
-            this.setState(function({ typedSearchQuery }){
-                if (query !== typedSearchQuery){
-                    return { 'typedSearchQuery' : query };
+            this.setState(function ({ typedSearchQuery }) {
+                if (query !== typedSearchQuery) {
+                    return { 'typedSearchQuery': query };
                 }
                 return null;
             });
         }
     }
 
-    toggleSearchAllItems(willSearchAllItems = !this.state.searchAllItems){
-        this.setState(function({ searchAllItems }){
-            if (typeof willSearchAllItems === 'boolean'){
-                if (willSearchAllItems === searchAllItems){
-                    return null;
+    getCurrentResultsSearchQuery(hrefParts) {
+        if (!hrefParts) {
+            hrefParts = url.parse(this.props.href, true);
+        }
+        return (hrefParts && hrefParts.query && hrefParts.query.q) || null;
+    }
+
+    getDropdownTitleText() {
+        const { href } = this.props;
+        const { searchItemType } = this.state;
+
+        if (!SearchBar.isBrowseOrSearchPage(href) && searchItemType === 'within') {
+            this.setState({ 'searchItemType': 'sets' });
```

Comment:
General practice/pattern is to **never** set state directly from a render method (`getDropdownTitleText()` is ultimately called from SearchBar's render method via `SelectItemTypeDropdownBtn` functional component). This is how we get infinite loops, pages/browsers that hang and can't be closed, etc.

Current version(s) of React thankfully don't allow this and block `setState` calls from being performed and instead print error message:
![](https://i.gyazo.com/38a3b883732cedc7324b1ebead9bf11e.png)

Instead, consider using [`getDerivedStateFromProps`](SelectItemTypeDropdownBtn) or `componentDidUpdate` or similar React lifecycle method for this.

Reply:
Oo! we'll fix it asap. Hanging browsers is the last thing we want.

Reply:
I don't think it would've happened in this case (if React didn't prevent) but in general / previously is common cause of issues. :-P

Somewhat tangentially (but not too relevant if using getDerivedStateFromProps or something), as of React v16, `setState` calls can be batched (if many occurring within short period of time) so the value of a `state` property isn't guaranteed to be accurate at time of calling setState. This is why React suggests to use functional state updater (`setState(function(currentState){ return { new state } })`) instead of plain setState w. obj `setState({ new state })` whenever a state change depends on an existing state value because they guarantee `currentState` is most up to date (it might be immediately returned/merged from another setState call and passed directly into this setState function call if multiple `setState`s are being batched/executed).

Reply:
@alexkb0009, we pushed a new commit and moved setState to componentDidUpdate. No warning and running well.